refactor: simplify proposal builder step management

adamgall opened the pull request:
Closes https://linear.app/decent-labs/issue/ENG-186/remove-url-page-state-from-proposal-builder

refactor: simplify proposal builder step management

- Remove URL path params for managing proposal builder steps
- Replace URL-based navigation with local React state in ProposalBuilder
- Update StepButtons component to use props instead of reading URL
- Rename contentRoute prop to mainContent for clarity
- Remove redundant redirects and path wildcards from router
- Update all proposal creation pages to use new step management pattern

This change simplifies the step management in the proposal creation flow by 
using React state instead of URL parameters. This makes the code more 
maintainable and removes unnecessary URL complexity while keeping the same 
functionality.

mudrila commented on src/components/ProposalBuilder/ProposalBuilder.tsx:
```diff
@@ -180,6 +176,12 @@ export function ProposalBuilder({
           !!formikProps.errors.nonce ||
           pendingCreateTx;
 
+        const { metadataStepButtons, transactionsStepButtons } = stepButtons({
```
I'd suggest passing `stepButtons` directly into `<StepButtons` as a property (maybe naming that something like `renderButtons`. The point would be to memoize result of rendering - current implementation will lead to calling `stepButtons` each time anything in the form will change, leading to a good bunch of unnecessary re-renderings

adamgall replied:
updated (I think?... can you check if the new code is what you were envisioning?)

mudrila commented on src/components/ProposalBuilder/ProposalBuilder.tsx:
```diff
@@ -212,33 +214,21 @@ export function ProposalBuilder({
-                              replace
-                            />
-                          }
+                      {currentStep === CreateProposalSteps.METADATA && (
```
I'd prefer if/else as it makes it more obvious that there're just 2 steps without going into `CreateProposalSteps` - better observability over expected rendering results without the need of getting additional component / types context

DarksightKellar replied:
+1

adamgall replied:
updated

mudrila commented on src/pages/dao/proposal-templates/new/SafeCreateProposalTemplatePage.tsx:
```diff
@@ -103,26 +99,26 @@ export function SafeCreateProposalTemplatePage() {
+      ),
+      transactionsStepButtons: (
+        <>
+          <PreviousButton onStepChange={onStepChange} />
```
Wondering if that makes sense to move `<PreviousButton` rendering right into `<StepButtons` - we know it will always be rendered on `transactions` step, will always have same behavior, so currently we just keep duplicating that in 3/4 usages, where in usage 4 it's just not gonna appear as there's no way to get into `transactions` step (but eventually there will be, so we'll need the same damn previous button). WDYT?

adamgall replied:
Something definitely needs fixed with these "step buttons".... i'll keep working on it

adamgall replied:
updated

mudrila commented on src/pages/dao/proposal-templates/new/SafeCreateProposalTemplatePage.tsx:
```diff
@@ -138,25 +134,21 @@ export function SafeCreateProposalTemplatePage() {
+              isProposalMode={true}
+              {...formikProps}
+            />
+            <ShowNonceInputOnMultisig
```
So the same stuff here - I think we're always expecting to render `ShowNonceInputOnMultisig` (at least, imo - we should be) - do we wanna put that alongside of rendering `mainContent` instead of just duplicating it over and over?

adamgall replied:
Oh yeah good call

adamgall replied:
updated

mudrila commented on src/pages/dao/proposals/new/sablier/SafeSablierProposalCreatePage.tsx:
```diff
@@ -149,26 +145,26 @@ export function SafeSablierProposalCreatePage() {
+      transactionsStepButtons: (
+        <>
+          <PreviousButton onStepChange={onStepChange} />
+          <CreateProposalButton isDisabled={createProposalBlocked} />
```
Again - `CreateProposalButton` is supposed to always occur on `transactions` step, so currently we keep duplicating it's rendering - maybe we should include that right into `StepButtons` and find a way to elegantly pass `createProposalBlocked` instead

adamgall replied:
updated

mudrila commented on src/pages/dao/proposals/new/sablier/SafeSablierProposalCreatePage.tsx:
```diff
@@ -184,24 +180,20 @@ export function SafeSablierProposalCreatePage() {
       prepareProposalData={prepareProposalData}
-      contentRoute={(formikProps, pendingCreateTx, nonce) => {
+      mainContent={(formikProps, pendingCreateTx, nonce, currentStep) => {
+        if (currentStep !== CreateProposalSteps.STREAMS) return null;
```
Should be removed after merging target branch - I killed this thingy

adamgall replied:
@mudrila Should this line be deleted entirely or changed to `CreateProposalSteps.TRANSACTIONS`?

mudrila commented on src/router.tsx:
```diff
@@ -164,45 +155,16 @@ export const router = (addressPrefix: string, daoAddress: string | undefined) =>
                 },
                 {
-                  path: 'new/*',
+                  path: 'new',
```
Oh hell yeah - now I wanna do the same for DAO Creator 🤣

mudrila reviewed (changes requested):
I feel like we have too much duplication around `StepButtons` and there're some performance problems with unnecessary re-renderings. Otherwise - looks great!

Da-Colon reviewed (approved):
No additional comments. LGTM

adamgall commented:
Marking this as `Draft` until I get back to it and clean some stuff up.

adamgall commented on src/components/ProposalBuilder/StepButtons.tsx:
```diff
@@ -1,55 +1,51 @@
+  onStepChange: (step: CreateProposalSteps) => void;
+}
+
+export function NextButton({ isDisabled, onStepChange }: StepButtonBaseProps) {
```
good idea! updated

adamgall commented:
@DarksightKellar @mudrila @Da-Colon @johnqh Moved this out of `draft`, ready for review again. I've addressed all of the review comments, and I think it's looking better.

Da-Colon reviewed (approved):
looks like it is working pretty smoothly.

![image](https://github.com/user-attachments/assets/b40ce9a3-c344-414f-b4ea-951045f458a0)

adamgall commented:
@mudrila do you still have changes requested on this PR? i'm being blocked by a previous review of yours.

adamgall commented:
@mudrila do you still have changes requested on this PR? i'm being blocked by a previous review of yours.
